alws/crud/errata.py

```diff
@@ -1109,44 +1087,207 @@ async def release_errata_record(db: AsyncSession, record_id: str):
+                )
+        release_tasks.append(
             release_errata_packages(
-                db,
+                session,
                 pulp,
                 db_record,
                 packages,
                 db_record.platform,
                 repo_href,
+                publish=publish,
             )
         )
-        publish_tasks.append(pulp.create_rpm_publication(repo_href))
-    await asyncio.gather(*tasks)
+        if publish:
+            publish_tasks.append(pulp.create_rpm_publication(repo_href))
+    if not publish:
+        return release_tasks
+    await asyncio.gather(*release_tasks)
     await asyncio.gather(*publish_tasks)
-    db_record.release_status = ErrataReleaseStatus.RELEASED
-    db_record.last_release_log = "Succesfully released"
-    await db.commit()
+
+
+def generate_query_for_release(records_ids: List[str]):
+    query = (
+        select(models.ErrataRecord)
+        .where(models.ErrataRecord.id.in_(records_ids))
+        .options(
+            selectinload(models.ErrataRecord.references),
+            selectinload(models.ErrataRecord.packages)
+            .selectinload(models.ErrataPackage.albs_packages)
+            .selectinload(models.ErrataToALBSPackage.build_artifact),
+            selectinload(models.ErrataRecord.platform)
+            .selectinload(models.Platform.repos),
+        )
+        .with_for_update()
+    )
+    return query
+
+
+async def release_errata_record(record_id: str):
+    pulp = PulpClient(
+        settings.pulp_host,
+        settings.pulp_user,
+        settings.pulp_password,
+    )
+    async with asynccontextmanager(get_db)() as session:
+        await session.execute(
+            update(models.ErrataRecord)
+            .where(models.ErrataRecord.id == record_id)
+            .values(
+                release_status=ErrataReleaseStatus.IN_PROGRESS,
+                last_release_log=None,
+            )
+        )
+        await session.commit()
+    async with asynccontextmanager(get_db)() as session:
+        session: AsyncSession
+        db_record = await session.execute(
+            generate_query_for_release([record_id]),
+        )
+        db_record: Optional[models.ErrataRecord] = db_record.scalars().first()
+        if not db_record:
+            logging.info("Record with %s id doesn't exists", record_id)
+            return
+
+        logging.info("Record release %s has been started", record_id)
+        search_params = prepare_search_params(db_record)
+        pulp_packages = await load_platform_packages(
+            db_record.platform,
+            search_params,
+            pulp,
+            for_release=True,
+        )
+        try:
+            repo_mapping = get_albs_packages_from_record(db_record, pulp_packages)
+        except Exception as exc:
+            db_record.release_status = ErrataReleaseStatus.FAILED
+            db_record.last_release_log = str(exc)
+            logging.exception("Cannot release %s record:", record_id)
```

Comment:
Not sure but, shall we log the error too? (in addition to store it as `last_release_log`)
Don't have a strong opinion, but if you ask me, I'd say yes, but I'd be okay if you don't think it's worth doing it ;)

Reply:
Nevermind, it's an exception and we'll somehow see what's going on behind the curtains